File: src/main/java/org/pyjjs/scheduler/core/logic/objectiveFunctions/ObjectiveFunction.java
package org.pyjjs.scheduler.core.logic.objectiveFunctions;

public abstract class ObjectiveFunction<T> {

    public abstract double calculate(T source, Context context);

    public double calculateAbsoluteDifference(T source, Context context) {
        return Math.abs(calculate(source, context) - getIdealValue());
    }

    public double getIdealValue() {
        return 0;
    }

}

File: src/main/kotlin/org/pyjjs/scheduler/core/scheduling/logic/objectiveFunctions/ObjectiveFunction.kt
package org.pyjjs.scheduler.core.scheduling.objectiveFunctions

abstract class ObjectiveFunction<T> {

    abstract fun calculate(source: T, context: Context): Double

    fun calculateAbsoluteDifference(source: T, context: Context): Double {
        return Math.abs(calculate(source, context) - idealValue)
    }

    val idealValue: Double
        get() = 0.0

}
